### PR TITLE
Add fourth Zephyr blog post

### PR DESCRIPTION
**Pull request by scottslowe**
This PR adds a Zephyr blog post about using stack references to connect multiple Pulumi projects.

### CODE REVIEW

**Comment by scottslowe** [comment]
Closes https://github.com/pulumi/devrel-team/issues/526 when merged

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:20]:
```suggestion
Here are links to all the blog posts in the series (entries below that are not linked are planned but haven't yet been published and will be updated as new posts are published):
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:33]:
```suggestion
This post is a continuation of the previous post in the series: Zephyr's infrastructure footprint and Pulumi codebase remains the same, that is, Zephyr now has _three_ Pulumi projects (zephyr-infra, zephyr-k8s, and zephyr-app). Each of these projects handles a different aspect of the infrastructure for Zephyr's online store:
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:47]:
```suggestion
Stack references are a feature in Pulumi that allow you to access output values from one stack in another stack. In other words, stack references provide a way to programmatically "link" stacks in different projects for the purpose of sharing information. You can find more details on stack references [in the Pulumi docs](/docs/intro/concepts/stack/#stackreferences).
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:51]:
```suggestion
* The code handling the Kubernetes cluster can reference the VPC ID and subnet IDs from the base infrastructure layer.
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:52]:
```suggestion
* Similarly, the code that is responsible for deploying the online store services can reference the necessary Kubernetes connection details (kubeconfig) for the Kubernetes cluster.
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:54]: We should either omit "generally" or explain the exceptions here. (I believe it never has an effect, but maybe there's something I'm not remembering.)

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:56]:
```suggestion
Additionally, any value retrieved via a stack reference is treated as a Pulumi Output, and therefore may require some additional work to transform values (such as the use of `Output.apply`). The recent addition of `OutputDetails` support in Pulumi---you can read more about that [here](/blog/stack-reference-output-details/)---helps enormously in this situation. Some SDKs also have language-specific mechanisms that can help; for example, using Go's `.AsStringArrayOutput()` method on a `StackReference.GetOutput` statement makes referencing subnet IDs from other project much easier.
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:58]:
```suggestion
While stack references are conceptually straightforward and easy to implement, there are some recommended practices to be mindful of regarding the use of stack outputs and stack references:
```

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (jkodroff) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:63]: I can't think of a use case where organization would need to be templates, but stack name definitely should be. It might be nice to show a default value that is the current stack name, e.g.

```
const stack = new pulumi.Config().getStack().split("/")[2];
const stackRef = new pulumi.StackReference(`my-organization/workshop-aws-proton-base-infra/${stack}`);
```

Double check the above code in an actual program before publishing - I can't vouch for its correctness, but it's something like that.

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:18]: Which portions, in particular, create that impression? (That will help me target the rewording appropriately.)

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:35]: It's too late to rename the repo now, but it's not to say that we can't show a justification for renaming the project (and the repo) later (this is a supported operation, after all). As for why the 3 repos/projects...well, that was the point of the previous post. :-) I can briefly summarize that again here just for the sake of completeness.

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:61]: @ringods @pierskarsenbarg Since you folks brought up the concern regarding minimize stack references and calls to read values, can you confirm if the concern is in stack outputs or stack references? Based on the comments in Slack, I believe it is the reading stack (and therefore stack references are the concern).

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:62]: @ringods @pierskarsenbarg Pinging you two again here. I read the exchange in Slack as needing to build an interface in the reading stack, but I could certainly see the need to build an interface/struct in the source stack as well. Can you confirm based on your understanding?

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:63]: To me, it seems entirely plausible that a company could have two organizations, with users that are members of both organizations, and there be a need to read information across the organizational boundary via a stack reference (two different business units, perhaps?).

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:65]: I'd really like readers to be able to see the highlighted code in the context of the broader Pulumi program. How strongly do you feel about putting code inline into the blog post?

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Addressed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Incorporated in f23ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:35]: Fixed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Partially addressed in f28ec13c. My English teacher wife and I agree that "allows" is correct, but I'll likely reword this in a future change just to avoid any confusion or disagreement. :-)

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Addressed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Fixed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Addressed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Fixed in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Done, taken care of in f28ec13c

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Handled by f28ec13c

**Comment by scottslowe** [comment]
@jkodroff Thanks for the detailed feedback. I believe I've addressed most of your points; there are still a couple outstanding, but I'll continue to work on them (I've responded to your comments, so if you have additional feedback in those areas I'd love to hear).

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (cnunciato) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:62]: >If it needs to be accessed from outside the stack, export it; otherwise, don't.

It occurred to me that after reading this, I might wonder, Why not? Is there any harm in doing so? There probably isn't, unless what you're exporting is somehow potentially sensitive of course. The "why" here might therefore be something along the lines of the same "why" that encourages encapsulation in a public API: expose what the API user cares about or is likely to need, hide all the rest as implementation detail (which ultimately leaves you with more ability to change the underlying implementation without affecting the API).

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (cnunciato) [inline themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:64]: I might tweak this slightly to clarify we're talking about the exporting, not the consuming (but up to you -- feel free to ignore if you don't think it adds anything):

```suggestion
3. **If you need to expose lots of outputs on a stack, consider exporting a structured data object.** It's possible to construct a JSON object (or dict or struct, depending on your language) to hold all the stack outputs, and then export that object. Then a stack reference can read that object, resulting in a single call to the backend. Be aware, though, that you'll need to write the necessary code to understand/import/unmarshall that JSON object in the referring stack, so there is a small amount of extra work required in this situation.
```

It might also be worth mentioning that this approach can also make it a little easier to codify the API contract between stacks and their consumers (vs. that contract consisting of "a bunch of optional strings"). 

Relatedly, we might want to mention that outputs are expected to be primitive/serializable types -- string, bool, object, etc. Occasionally folks think they can export whole resource instances (as you would in a typical program) and then realize this doesn't quite work, and are confused. (The docs may also mention this, though.)

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (cnunciato) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:65]: +1 to add some code here to show what "parameterize your stack references" looks like.

**Review by cnunciato — COMMENTED** [review]
Looking good! Some suggestions for you to consider.

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (susanev) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: we also have this in our style guide https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:77]: > I link to link these references
To which references are you referring? The blog post references? Or the GitHub references? (I think the latter, if I'm reading your comment correctly.)

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Reworded completely in 17aacd6f

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Code added in 17aacd6f, along with references to the code in other parts of the post

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:69]: Code samples added in 17aacd6f

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Fixed in 17aacd6f

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:54]: Code was added toward the end in 17aacd6f and I've put a reference to that code here

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Done, fixed in 17aacd6f

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: I've added some text, borrowing from your suggestion, in 17aacd6f

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: In 17aacd6f I've reworded per your suggestion, but there is still an outstanding question as to whether it is the stack outputs that are "costly" or if it is the stack references reading those outputs that's "costly" (from a latency perspective). I've asked @ringods & @pierskarsenbarg to weigh in on that, as I'm not sure.

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (cnunciato) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:116]: You know what, I'm sorry -- I must've misread this. I think I read it as "The code above refers to the `multi-project` tag in the repository..." so my suggestion was just to link "multi-project tag" to the tag in the repo. :ignoreme:

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (cnunciato) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md:35]: I do think it would be nice to link these projects to their respective repos, though. What do you think? Should we not?

**Review by cnunciato — APPROVED** [review]
Nice! Somehow it feels like it ends a bit quickly (I was all into following those code snippets!), but that's okay -- if there's nothing more to be said, there's nothing more to be said! 😄 The snippets totally help, though -- thanks for adding those. 

Other than a couple of minor suggestions to add some links and pull in that sweet graphic you made showing the stacks and their relationships (and I guess adding a meta image as well before merging), this looks good to me. 👍

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: I think this has been resolved (refer to Slack discussions in #project-zephyr and #questions).

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: After some lengthy Slack discussions, I think this has been resolved.

File: themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-tying-stacks-together/index.md]: Resolved via Slack discussions (see threads in `#project-zephyr` and `#questions`).

File: themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md
Comment (ringods) [inline themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:71]: I would add a 6th bullet point talking about security. With separate stacks also comes that not everyone should be allowed to access the stack outputs of the production stack. I would highlight [Stack Permissions](https://www.pulumi.com/docs/intro/pulumi-service/projects-and-stacks/#stack-permissions) and [Teams and RBAC](https://www.pulumi.com/docs/intro/pulumi-service/teams/) to control such access.

**Review by ringods — CHANGES_REQUESTED** [review]
@scottslowe in the beginning of each article, you have the overview of all articles in the series. In the first 3 articles, you didn't activate the link to this 4th article yet.

File: themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md
Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:71]: This is an excellent suggestion as a revision to the previous post, which discusses _why_ you would use multiple projects. Unless I'm misreading your suggestion, it sounds like it would fall squarely in that post as opposed to this one (which is focused on the use of stack references).

Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:71]: I've created https://github.com/pulumi/devrel-team/issues/530 to track this work.

Comment (ringods) [reply themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:71]: I would integrate it here. After splitting the teams (article 3), you are now introducing the stack references. If you layer projects (infra, k8s, app), you can also talk about stack permissions explaining that the access to the infra and/or k8s is read-only to the app team, while read-write for the infra team.

If you export secrets in a stack, you definitely want to limit access to these, hence also my suggestion (in this article) to narrow the permissions overall to the production stacks of each Pulumi program.

Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:71]: Keep in mind that the next post in the series (after we take a quick break to look at local testing) is security and RBAC, in which we will most definitely look at this in more detail. I'll see if I can work in a quick mention here with a statement that we'll be exploring the topic in more detail in an upcoming post.

Comment (scottslowe) [reply themes/default/content/blog/iac-recommended-practices-using-stack-references/index.md:70]: Fixed in b34f9cd9

**Comment by scottslowe** [comment]
> @scottslowe in the beginning of each article, you have the overview of all articles in the series. In the first 3 articles, you didn't activate the link to this 4th article yet.

This is addressed in b34f9cd9, thanks!

**Comment by scottslowe** [comment]
Just waiting on final meta image before we can merge.

**Review by cnunciato — APPROVED** [review]
Nice work -- looks great. 🚀

**Pull request merged**
